// testcafe-hammerhead/lib/utils/promisified-functions.js
'use strict';

exports.__esModule = true;
exports.exec = exports.fsObjectExists = exports.writeFile = exports.makeDir = exports.readDir = exports.access = exports.stat = exports.readFile = exports.inflateRaw = exports.inflate = exports.gunzip = exports.deflate = exports.gzip = undefined;

var _zlib = require('zlib');

var _zlib2 = _interopRequireDefault(_zlib);

var _promisify = require('./promisify');

var _promisify2 = _interopRequireDefault(_promisify);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _child_process = require('child_process');

var _child_process2 = _interopRequireDefault(_child_process);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const gzip = exports.gzip = (0, _promisify2.default)(_zlib2.default.gzip);
const deflate = exports.deflate = (0, _promisify2.default)(_zlib2.default.deflate);
const gunzip = exports.gunzip = (0, _promisify2.default)(_zlib2.default.gunzip);
const inflate = exports.inflate = (0, _promisify2.default)(_zlib2.default.inflate);
const inflateRaw = exports.inflateRaw = (0, _promisify2.default)(_zlib2.default.inflateRaw);
const readFile = exports.readFile = (0, _promisify2.default)(_fs2.default.readFile);
const stat = exports.stat = (0, _promisify2.default)(_fs2.default.stat);
const access = exports.access = (0, _promisify2.default)(_fs2.default.access);
const readDir = exports.readDir = (0, _promisify2.default)(_fs2.default.readdir);
const makeDir = exports.makeDir = (0, _promisify2.default)(_fs2.default.mkdir);
const writeFile = exports.writeFile = (0, _promisify2.default)(_fs2.default.writeFile);
const fsObjectExists = exports.fsObjectExists = fsPath => stat(fsPath).then(() => true, () => false);
const exec = exports.exec = (0, _promisify2.default)(_child_process2.default.exec);
