// testcafe-hammerhead/lib/proxy/router.js
'use strict';

exports.__esModule = true;

var _http = require('../utils/http');

var _cryptoMd = require('crypto-md5');

var _cryptoMd2 = _interopRequireDefault(_cryptoMd);

var _url = require('../utils/url');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*eslint-enable no-unused-vars*/
const PARAM_RE = /^{(\S+)}$/;
function buildRouteParamsMap(routeMatch, paramNames) {
    return paramNames.reduce((params, paramName, i) => {
        params[paramName] = routeMatch[i + 1];
        return params;
    }, {});
}
class Router {
    constructor(options = {}) {
        this.routes = new Map();
        this.routesWithParams = [];
        this.options = options;
    }
    _registerRoute(route, method, handler) {
        const tokens = route.split('/');
        const isRouteWithParams = tokens.some(token => PARAM_RE.test(token));
        if (isRouteWithParams) {
            if (typeof handler === 'function') this._registerRouteWithParams(tokens, method, handler);
        } else {
            const routeName = `${method} ${route}`;
            if (typeof handler !== 'function') {
                this._processStaticContent(handler);
                handler.etag = (0, _cryptoMd2.default)(handler.content);
                this.routes.set(routeName, { handler, isStatic: true });
            } else this.routes.set(routeName, { handler, isStatic: false });
        }
    }
    _registerRouteWithParams(tokens, method, handler) {
        const paramNames = [];
        const reParts = tokens.map(token => {
            const paramMatch = token.match(PARAM_RE);
            if (paramMatch) {
                paramNames.push(paramMatch[1]);
                return '(\\S+?)';
            }
            return token;
        });
        this.routesWithParams.push({ handler, paramNames, re: new RegExp(`^${method} ${reParts.join('/')}$`) });
    }
    _route(req, res, serverInfo) {
        const routerQuery = `${req.method} ${(0, _url.getPathname)(req.url)}`;
        const route = this.routes.get(routerQuery);
        if (route) {
            if (route.isStatic) (0, _http.respondStatic)(req, res, route.handler, this.options.staticContentCaching);else route.handler(req, res, serverInfo); // eslint-disable-line no-extra-parens
            return true;
        }
        for (var _iterator = this.routesWithParams, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
            var _ref;

            if (_isArray) {
                if (_i >= _iterator.length) break;
                _ref = _iterator[_i++];
            } else {
                _i = _iterator.next();
                if (_i.done) break;
                _ref = _i.value;
            }

            const routeWithParams = _ref;

            const routeMatch = routerQuery.match(routeWithParams.re);
            if (routeMatch) {
                const params = buildRouteParamsMap(routeMatch, routeWithParams.paramNames);
                routeWithParams.handler(req, res, serverInfo, params);
                return true;
            }
        }
        return false;
    }
    // API
    GET(route, handler) {
        this._registerRoute(route, 'GET', handler);
    }
    POST(route, handler) {
        this._registerRoute(route, 'POST', handler);
    }
}
exports.default = Router;
module.exports = exports['default'];
