// testcafe/lib/compiler/test-file/base.js
'use strict';

exports.__esModule = true;

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _lodash = require('lodash');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class TestFileCompilerBase {
    constructor() {
        const escapedExt = (0, _lodash.escapeRegExp)(this.getSupportedExtension());

        this.supportedExtensionRe = new RegExp(`${escapedExt}$`);
    }

    _hasTests() /* code */{
        throw new Error('Not implemented');
    }

    getSupportedExtension() {
        throw new Error('Not implemented');
    }

    precompile() /* testFilesInfo */{
        return (0, _asyncToGenerator3.default)(function* () {
            throw new Error('Not implemented');
        })();
    }

    compile() /* code, filename */{
        return (0, _asyncToGenerator3.default)(function* () {
            throw new Error('Not implemented');
        })();
    }

    execute() /* compiledCode, filename */{
        return (0, _asyncToGenerator3.default)(function* () {
            throw new Error('Not implemented');
        })();
    }

    canCompile(code, filename) {
        return this.supportedExtensionRe.test(filename);
    }

    get canPrecompile() {
        return false;
    }

    cleanUp() {
        // NOTE: Optional. Do nothing by default.
    }
}
exports.default = TestFileCompilerBase;
module.exports = exports['default'];
