// testcafe/lib/errors/runtime/index.js
'use strict';

exports.__esModule = true;
exports.CompositeError = exports.ClientFunctionAPIError = exports.APIError = exports.TestCompilationError = exports.GeneralError = undefined;

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _callsiteRecord = require('callsite-record');

var _templates = require('./templates');

var _templates2 = _interopRequireDefault(_templates);

var _createStackFilter = require('../create-stack-filter');

var _createStackFilter2 = _interopRequireDefault(_createStackFilter);

var _getCallsite = require('../get-callsite');

var _renderTemplate = require('../../utils/render-template');

var _renderTemplate2 = _interopRequireDefault(_renderTemplate);

var _types = require('../types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ERROR_SEPARATOR = '\n\n';

class ProcessTemplateInstruction {
    constructor(processFn) {
        this.processFn = processFn;
    }
}

// Errors
class GeneralError extends Error {
    constructor(...args) {
        const code = args.shift();
        const template = _templates2.default[code];

        super((0, _renderTemplate2.default)(template, ...args));

        (0, _assign2.default)(this, { code, data: args });
        Error.captureStackTrace(this, GeneralError);
    }
}

exports.GeneralError = GeneralError;
class TestCompilationError extends Error {
    constructor(originalError) {
        const template = _templates2.default[_types.RUNTIME_ERRORS.cannotPrepareTestsDueToError];
        const errorMessage = originalError.toString();

        super((0, _renderTemplate2.default)(template, errorMessage));

        (0, _assign2.default)(this, {
            code: _types.RUNTIME_ERRORS.cannotPrepareTestsDueToError,
            data: [errorMessage]
        });

        // NOTE: stack includes message as well.
        this.stack = (0, _renderTemplate2.default)(template, originalError.stack);
    }
}

exports.TestCompilationError = TestCompilationError;
class APIError extends Error {
    constructor(methodName, code, ...args) {
        let template = _templates2.default[code];

        template = APIError._prepareTemplateAndArgsIfNecessary(template, args);

        const rawMessage = (0, _renderTemplate2.default)(template, ...args);

        super((0, _renderTemplate2.default)(_templates2.default[_types.RUNTIME_ERRORS.cannotPrepareTestsDueToError], rawMessage));

        (0, _assign2.default)(this, { code, data: args });

        // NOTE: `rawMessage` is used in error substitution if it occurs in test run.
        this.rawMessage = rawMessage;
        this.callsite = (0, _getCallsite.getCallsiteForMethod)(methodName);

        // NOTE: We need property getters here because callsite can be replaced by an external code.
        // See https://github.com/DevExpress/testcafe/blob/v1.0.0/src/compiler/test-file/formats/raw.js#L22
        // Also we can't use an ES6 getter for the 'stack' property, because it will create a getter on the class prototype
        // that cannot override the instance property created by the Error parent class.
        (0, _defineProperties2.default)(this, {
            'stack': {
                get: () => this._createStack(_callsiteRecord.renderers.noColor)
            },

            'coloredStack': {
                get: () => this._createStack(_callsiteRecord.renderers.default)
            }
        });
    }

    _renderCallsite(renderer) {
        if (!this.callsite) return '';

        // NOTE: Callsite will throw during rendering if it can't find a target file for the specified function or method:
        // https://github.com/inikulin/callsite-record/issues/2#issuecomment-223263941
        try {
            return this.callsite.renderSync({
                renderer: renderer,
                stackFilter: (0, _createStackFilter2.default)(Error.stackTraceLimit)
            });
        } catch (error) {
            return '';
        }
    }

    _createStack(renderer) {
        const renderedCallsite = this._renderCallsite(renderer);

        if (!renderedCallsite) return this.message;

        return this.message + ERROR_SEPARATOR + renderedCallsite;
    }

    static _prepareTemplateAndArgsIfNecessary(template, args) {
        const lastArg = args.pop();

        if (lastArg instanceof ProcessTemplateInstruction) template = lastArg.processFn(template);else args.push(lastArg);

        return template;
    }
}

exports.APIError = APIError;
class ClientFunctionAPIError extends APIError {
    constructor(methodName, instantiationCallsiteName, code, ...args) {
        args.push(new ProcessTemplateInstruction(template => template.replace(/\{#instantiationCallsiteName\}/g, instantiationCallsiteName)));

        super(methodName, code, ...args);
    }
}

exports.ClientFunctionAPIError = ClientFunctionAPIError;
class CompositeError extends Error {
    constructor(errors) {
        super(errors.map(({ message }) => message).join(ERROR_SEPARATOR));

        this.stack = errors.map(({ stack }) => stack).join(ERROR_SEPARATOR);
        this.code = _types.RUNTIME_ERRORS.compositeArgumentsError;
    }
}
exports.CompositeError = CompositeError;
