// testcafe/lib/test-run/commands/service.js
'use strict';

exports.__esModule = true;
exports.UnlockPageCommand = exports.BackupStoragesCommand = exports.TestDoneCommand = exports.SetBreakpointCommand = exports.HideAssertionRetriesStatusCommand = exports.ShowAssertionRetriesStatusCommand = undefined;

var _type = require('./type');

var _type2 = _interopRequireDefault(_type);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Commands
class ShowAssertionRetriesStatusCommand {
    constructor(timeout) {
        this.type = _type2.default.showAssertionRetriesStatus;
        this.timeout = timeout;
    }
}

exports.ShowAssertionRetriesStatusCommand = ShowAssertionRetriesStatusCommand;
class HideAssertionRetriesStatusCommand {
    constructor(success) {
        this.type = _type2.default.hideAssertionRetriesStatus;
        this.success = success;
    }
}

exports.HideAssertionRetriesStatusCommand = HideAssertionRetriesStatusCommand;
class SetBreakpointCommand {
    constructor(isTestError) {
        this.type = _type2.default.setBreakpoint;
        this.isTestError = isTestError;
    }
}

exports.SetBreakpointCommand = SetBreakpointCommand;
class TestDoneCommand {
    constructor() {
        this.type = _type2.default.testDone;
    }
}

exports.TestDoneCommand = TestDoneCommand;
class BackupStoragesCommand {
    constructor() {
        this.type = _type2.default.backupStorages;
    }
}

exports.BackupStoragesCommand = BackupStoragesCommand;
class UnlockPageCommand {
    constructor() {
        this.type = _type2.default.unlockPage;
    }
}
exports.UnlockPageCommand = UnlockPageCommand;
