// testcafe/lib/live/file-watcher/index.js
'use strict';

exports.__esModule = true;

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _modulesGraph = require('./modules-graph');

var _modulesGraph2 = _interopRequireDefault(_modulesGraph);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const WATCH_LOCKED_TIMEOUT = 700;

class FileWatcher extends _events2.default {
    constructor(files) {
        super();

        this.FILE_CHANGED_EVENT = 'file-changed';

        this.watchers = {};
        this.lockedFiles = {};
        this.modulesGraph = null;
        this.lastChangedFiles = [];

        files.forEach(f => this.addFile(f));
    }

    _onChanged(file) {
        const cache = require.cache;

        if (!this.modulesGraph) {
            this.modulesGraph = new _modulesGraph2.default();
            this.modulesGraph.build(cache, (0, _keys2.default)(this.watchers));
        } else {
            this.lastChangedFiles.forEach(changedFile => this.modulesGraph.rebuildNode(cache, changedFile));
            this.lastChangedFiles = [];
        }

        this.lastChangedFiles.push(file);
        this.modulesGraph.clearParentsCache(cache, file);

        this.emit(this.FILE_CHANGED_EVENT, { file });
    }

    _watch(file) {
        if (this.lockedFiles[file]) return;

        this.lockedFiles[file] = setTimeout(() => {
            this._onChanged(file);

            delete this.lockedFiles[file];
        }, WATCH_LOCKED_TIMEOUT);
    }

    stop() {
        (0, _values2.default)(this.watchers).forEach(watcher => {
            watcher.close();
        });
    }

    addFile(file) {
        if (!this.watchers[file] && file.indexOf('node_modules') < 0) {
            if (this.modulesGraph) {
                this.lastChangedFiles.push(file);
                this.modulesGraph.addNode(file, require.cache);
            }

            this.watchers[file] = _fs2.default.watch(file, () => this._watch(file));
        }
    }
}
exports.default = FileWatcher;
module.exports = exports['default'];
