// testcafe/lib/live/test-run.js
'use strict';

exports.__esModule = true;
exports.TestRunCtorFactory = undefined;

var _pinkie = require('pinkie');

var _pinkie2 = _interopRequireDefault(_pinkie);

var _testRun = require('../test-run');

var _testRun2 = _interopRequireDefault(_testRun);

var _testRunState = require('./test-run-state');

var _testRunState2 = _interopRequireDefault(_testRunState);

var _type = require('../test-run/commands/type');

var _type2 = _interopRequireDefault(_type);

var _service = require('../test-run/commands/service');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const TEST_RUN_ABORTED_MESSAGE = 'The test run has been aborted.';

const TestRunCtorFactory = exports.TestRunCtorFactory = function TestRunCtorFactory(callbacks) {
    const created = callbacks.created,
          done = callbacks.done,
          readyToNext = callbacks.readyToNext;


    return class LiveModeTestRun extends _testRun2.default {
        constructor(test, browserConnection, screenshotCapturer, warningLog, opts) {
            super(test, browserConnection, screenshotCapturer, warningLog, opts);

            created(this, test);

            this.state = _testRunState2.default.created;
            this.finish = null;
            this.stopping = false;
            this.isInRoleInitializing = false;
            this.stopped = false;
        }

        stop() {
            this.stopped = true;
        }

        _useRole(...args) {
            this.isInRoleInitializing = true;

            return super._useRole.apply(this, args).then(res => {
                this.isInRoleInitializing = false;

                return res;
            }).catch(err => {
                this.isInRoleInitializing = false;

                throw err;
            });
        }

        executeCommand(commandToExec, callsite, forced) {
            // NOTE: don't close the page and the session when the last test in the queue is done
            if (commandToExec.type === _type2.default.testDone && !forced) {
                done(this, this.stopped).then(() => this.executeCommand(commandToExec, callsite, true)).then(() => readyToNext(this));

                this.executeCommand(new _service.UnlockPageCommand(), null);

                return _pinkie2.default.resolve();
            }

            if (this.stopped && !this.stopping && !this.isInRoleInitializing) {
                this.stopping = true;

                return _pinkie2.default.reject(new Error(TEST_RUN_ABORTED_MESSAGE));
            }

            return super.executeCommand(commandToExec, callsite);
        }
    };
};
