// testcafe/lib/live/test-run-controller.js
'use strict';

exports.__esModule = true;

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _pinkie = require('pinkie');

var _pinkie2 = _interopRequireDefault(_pinkie);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _testRun = require('./test-run');

var _testRunState = require('./test-run-state');

var _testRunState2 = _interopRequireDefault(_testRunState);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LiveModeTestRunController extends _events2.default {
    constructor() {
        super();

        this.testWrappers = [];
        this.expectedTestCount = 0;
        this._testRunCtor = null;

        this.testRuns = {};
        this.allTestsCompletePromise = _pinkie2.default.resolve();
        this.completeAllRunningTests = _lodash2.default;

        this.on('all-tests-complete', () => this.completeAllRunningTests());
    }

    get TestRunCtor() {
        if (!this._testRunCtor) {
            this._testRunCtor = (0, _testRun.TestRunCtorFactory)({
                created: testRun => this._onTestRunCreated(testRun),
                done: (testRun, forced) => this._onTestRunDone(testRun, forced),
                readyToNext: testRun => this._onTestRunReadyToNext(testRun)
            });
        }

        return this._testRunCtor;
    }

    setExpectedTestCount(testCount) {
        this.expectedTestCount = testCount;
    }

    _getTestRuns() {
        return [].concat(...(0, _values2.default)(this.testRuns));
    }

    run() {
        const readyToNextPromises = [];

        const testRuns = [].concat(...(0, _values2.default)(this.testRuns));

        testRuns.forEach(testRun => {
            if (testRun.finish) {
                readyToNextPromises.push(testRun.readyToNextPromise);
                testRun.finish();
            }
        });

        this.testRuns = {};

        return _pinkie2.default.all(readyToNextPromises);
    }

    stop() {
        this._getTestRuns().forEach(testRun => {
            testRun.stop();
        });
    }

    _getTestWrapper(test) {
        return this.testWrappers.find(w => w.test === test);
    }

    _onTestRunCreated(testRun) {

        this.allTestsCompletePromise = new _pinkie2.default(resolve => {
            this.completeAllRunningTests = resolve;
        });

        const connectionId = testRun.browserConnection.id;

        this.testRuns[connectionId] = this.testRuns[connectionId] || [];

        this.testRuns[connectionId].push(testRun);
    }

    _onTestRunDone(testRun) {
        testRun.state = _testRunState2.default.done;

        const hasRunningTests = this._getTestRuns().some(t => t.state !== _testRunState2.default.done);

        if (!hasRunningTests) this.emit('all-tests-complete');

        const browserTestRuns = this.testRuns[testRun.browserConnection.id];

        testRun.readyToNextPromise = new _pinkie2.default(resolve => {
            testRun.setReadyToNext = resolve;
        });

        if (browserTestRuns.length < this.expectedTestCount || browserTestRuns.some(t => t.state !== _testRunState2.default.done)) return _pinkie2.default.resolve();

        return new _pinkie2.default(resolve => {
            testRun.finish = () => {
                testRun.finish = null;
                testRun.state = _testRunState2.default.done;
                resolve();
            };
        });
    }

    _onTestRunReadyToNext(testRun) {
        testRun.setReadyToNext();
    }
}

exports.default = LiveModeTestRunController;
module.exports = exports['default'];
