// testcafe/lib/embedding-utils.js
'use strict';

exports.__esModule = true;
const lazyRequire = require('import-lazy')(require);
const hammerhead = lazyRequire('testcafe-hammerhead');
const ReporterPluginHost = lazyRequire('./reporter/plugin-host');
const TestRunErrorFormattableAdapter = lazyRequire('./errors/test-run/formattable-adapter');
const testRunErrors = lazyRequire('./errors/test-run');
const COMMAND_TYPE = lazyRequire('./test-run/commands/type');
const getTestListModule = lazyRequire('./compiler/test-file/formats/es-next/get-test-list');
const getTypeScriptTestListModule = lazyRequire('./compiler/test-file/formats/typescript/get-test-list');
const getCoffeeScriptTestListModule = lazyRequire('./compiler/test-file/formats/coffeescript/get-test-list');
const initializers = lazyRequire('./test-run/commands/validations/initializers');
const errorTypes = lazyRequire('./error/types');
const createCommandFromObject = lazyRequire('./test-run/commands/from-object');
const processTestFnError = lazyRequire('./errors/process-test-fn-error');

// NOTE: we can't use lazy require for TestRun and Assignable, because it breaks prototype chain for inherited classes
let TestRun = null;
let Assignable = null;

exports.default = {
    TestRunErrorFormattableAdapter,
    testRunErrors,
    COMMAND_TYPE,
    errorTypes,

    get Assignable() {
        if (!Assignable) Assignable = require('./utils/assignable');

        return Assignable;
    },

    get TestRun() {
        if (!TestRun) TestRun = require('./test-run');

        return TestRun;
    },

    get getTestList() {
        return getTestListModule.getTestList;
    },

    get getTypeScriptTestList() {
        return getTypeScriptTestListModule.getTypeScriptTestList;
    },

    get getCoffeeScriptTestList() {
        return getCoffeeScriptTestListModule.getCoffeeScriptTestList;
    },

    get getTestListFromCode() {
        return getTestListModule.getTestListFromCode;
    },

    get getTypeScriptTestListFromCode() {
        return getTypeScriptTestListModule.getTypeScriptTestListFromCode;
    },

    get getCoffeeScriptTestListFromCode() {
        return getCoffeeScriptTestListModule.getCoffeeScriptTestListFromCode;
    },

    get initSelector() {
        return initializers.initSelector;
    },

    get createCommandFromObject() {
        return createCommandFromObject;
    },

    get processTestFnError() {
        return processTestFnError;
    },

    ensureUploadDirectory(...args) {
        return hammerhead.UploadStorage.ensureUploadsRoot(...args);
    },

    copyFilesToUploadFolder(...args) {
        return hammerhead.UploadStorage.copy(...args);
    },

    buildReporterPlugin(pluginFactory, outStream) {
        const plugin = pluginFactory();

        return new ReporterPluginHost(plugin, outStream);
    }
};
module.exports = exports['default'];
