// testcafe/lib/utils/get-filter-fn.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

exports.default = function (opts) {
    const filteringOpts = (0, _lodash.pick)(opts, (0, _keys2.default)(FILTERING_OPTIONS));

    if (isAllFilteringOptionsAreUndefined(filteringOpts)) return void 0;

    return (0, _assign2.default)(createFilterFn(filteringOpts), filteringOpts);
};

var _lodash = require('lodash');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const FILTERING_OPTIONS = {
    testGrep: 'testGrep',
    fixtureGrep: 'fixtureGrep',
    testMeta: 'testMeta',
    fixtureMeta: 'fixtureMeta',
    test: 'test',
    fixture: 'fixture'
};

function isAllFilteringOptionsAreUndefined(opts) {
    return (0, _keys2.default)(FILTERING_OPTIONS).every(option => opts[option] === void 0);
}

function createFilterFn(opts) {
    return (testName, fixtureName, fixturePath, testMeta, fixtureMeta) => {
        if (opts.test && testName !== opts.test) return false;

        if (opts.testGrep && !opts.testGrep.test(testName)) return false;

        if (opts.fixture && fixtureName !== opts.fixture) return false;

        if (opts.fixtureGrep && !opts.fixtureGrep.test(fixtureName)) return false;

        if (opts.testMeta && !(0, _lodash.isMatch)(testMeta, opts.testMeta)) return false;

        if (opts.fixtureMeta && !(0, _lodash.isMatch)(fixtureMeta, opts.fixtureMeta)) return false;

        return true;
    };
}

module.exports = exports['default'];
